src/Storage/ClaimController.ts

```diff
@@ -53,51 +54,146 @@ export class ClaimController {
     ])
+    logger.trace('finished claim update has pairs')
+    return {
+      claim,
+      entry,
+      ...rest,
+    }
+  }
 
+  async publishEntryDownload({ entry, claim, ...rest }: { claim: Claim; entry: Entry }) {
+    const logger = this.logger.child({ method: 'publishEntryDownload' })
+    logger.trace('started claim publishing')
     await this.messaging.publishClaimsDownloaded([
       {
         claim,
-        ipfsHash,
+        ipfsHash: entry.ipfsHash,
       },
     ])
+    logger.trace('finished claim publishing')
+    return {
+      claim,
+      entry,
+      ...rest,
+    }
+  }
+
+  async downloadEntryClaim({ entry, ...rest }: { entry: Entry }) {
+    const logger = this.logger.child({ method: 'downloadEntryClaim' })
+    logger.trace('starting claim download')
+    const claim = await this.downloadClaim(entry.ipfsHash)
+    logger.trace('finished claim download', claim)
+    return {
+      entry,
+      claim,
+      ...rest,
+    }
+  }
+
+  async findEntryToDownload({
+    currentTime = new Date().getTime(),
+    retryDelay,
+    maxAttempts,
+    ...rest
+  }: {
+    currentTime?: number
+    retryDelay: number
+    maxAttempts: number
+  }) {
+    const logger = this.logger.child({ method: 'findEntryToDownload' })
+    logger.info('started finding claim')
+    const entry = await this.collection.findOne({
+      claimId: null,
+      ipfsHash: { $exists: true },
+      $and: [
+        {
+          $or: [
+            { lastDownloadAttempt: null },
+            { lastDownloadAttempt: { $exists: false } },
+            { lastDownloadAttempt: { $lt: currentTime - retryDelay } },
+          ],
+        },
+        {
+          $or: [{ lastDownloadSuccess: null }, { lastDownloadSuccess: { $exists: false } }],
+        },
+        {
+          $or: [
+            { downloadAttempts: null },
+            { downloadAttempts: { $exists: false } },
+            { downloadAttempts: { $lte: maxAttempts } },
+          ],
+        },
+      ],
+    })
+    logger.info('finished finding claim', entry)
+    return {
+      currentTime,
+      retryDelay,
+      maxAttempts,
+      entry,
+      ...rest,
+    }
+  }
+
+  async updateEntryAttempts({
+    entry,
+    currentTime = new Date().getTime(),
+    ...rest
+  }: {
+    entry: Entry
+    currentTime?: number
+  }) {
+    const logger = this.logger.child({ method: 'updateEntryAttempts' })
+    logger.trace('started updating claim')
+    await this.collection.updateOne(
+      {
+        _id: entry._id,
+      },
+      {
+        $set: { lastDownloadAttempt: currentTime },
+        $inc: { downloadAttempts: 1 },
+      }
+    )
+    logger.trace('finished updating claim')
+    return {
+      entry,
+      currentTime,
+      ...rest,
+    }
+  }
+
+  async downloadNextHash({ retryDelay = 600000, maxAttempts = 20 }: { retryDelay: number; maxAttempts: number }) {
+    return this.findEntryToDownload({ retryDelay, maxAttempts })
+      .then(this.updateEntryAttempts.bind(this))
```

Comment:
I am not accustomed to using classes, I need to see if I can structure this in a way that doesn't require bind.

Reply:
I originally wanted to use ramda pipeP, but that doesn't seem to have typescript types.

Reply:
Yep, we want to avoid that pattern. If you define the functions as arrow functions you don't need to use bind. 

Regarding the `.then`'s, I prefer to avoid using promises directly completely, always using async/await. 

But if you prefer to use `.then` to chain calls, you can remove the `async` from the function declaration since it's not needed.

Reply:
Yeah, I was trying to follow current coding style while also creating a composable flow that avoided a bunch of try/catch everywhere. I was thinking about removing the async from the functions and doing arrow functions but didn't want to break the coding style that much.